### PR TITLE
CLI tool and GH workflow to highlight PR changes

### LINKED ISSUE #29814: GH action for adding label after if a PR introduces new filter/action, or if there are template changes.
We highlight this information in our release post, however, it's time-consuming to compile this information. Automatically adding labels to PRs which are modifying filters, actions, and template changes will save time every release.

False positives are fine, for example, when there are changes in action even though it already exists, and release lead can remove those labels while going through the PR.

### PR DESCRIPTION
**Pull request by claudiosanches**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

CLI tool to analyze PRs for:

- Templates updated
- New actions and filters

GH workflow:

- Runs this CLI tool and insert a comment with all results

This tool will check for the plugin version and compare against hooks and templates changes found, to make sure all are new changes introduced in the latest release.

Fixes https://github.com/woocommerce/woocommerce/issues/29814

### How to test the changes in this Pull Request:

1. Check out this branch
2. Run `pnpm install`
3. Create a new branch and introduce some new hook with a `@since (make sure is the same as your `woocommerce.php`)
4. Make changes in some templates, and in one of the templates you changed make an update to the `@version` docblock (make sure is the same as your `woocommerce.php`).
5. Commit and push your branch to GH.
6. Run `./tools/code-analyzer/bin/dev analyzer BRANCH_TO_COMPARE` 
7. Fork this repository and merge this PR into trunk on your forked repository.
8. Edit line 18 of `.github/workflows/pr-highlight-changes.yml`, to include your forked repo `./tools/code-analyzer/bin/dev analyzer "$GITHUB_HEAD_REF" -o github `-s <github-org/repo>` (replace with your forked repo slug).
9. Go to the actions tab and enable it.
10. Go to your repo settings and set a GH token to make the commits in your forked repository using the `GITHUB_HEAD_REF` ID.
11. Create a new PR in your forked repository including some new hooks, and making template changes like in step `4`.
12. Check if the workflow is running correctly and if includes comments to the PR.

```
./tools/code-analyzer/bin/dev analyzer add/changes -s claudiosanches/woocommerce 
```

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

None

### FOR PR REVIEWER ONLY:

* [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

**Comment by roykho** [comment]
I think we need to add the tool to this workspace https://github.com/woocommerce/woocommerce/blob/trunk/pnpm-workspace.yaml otherwise `pnpm install` will not install the modules.

I tried adding some new hooks but it didn't seem to detect it? I did something like this:
```
/**
 * Test hook
 *
 * @since 6.5
 */
do_action( 'woocommerce_test_hook' );
```

**Comment by claudiosanches** [comment]
@roykho the regex uses the full version number, that's why doesn't match with `6.5`, it is search for `6.5.0`. I'll try to improve it.

File: tools/code-analyzer/bin/run.cmd [hunk 1]
```diff
@@ -0,0 +1,3 @@
+@echo off
+
+node "%~dp0\run" %*
```
Comment (roykho) [inline tools/code-analyzer/bin/run.cmd]: Need an empty line at EOF.

File: tools/code-analyzer/src/commands/analyzer/index.ts [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+
+		if (!content.match(/diff --git a\/(.+).php/g)) {
+			CliUx.ux.action.stop();
+			return report;
+		}
+
+		const matchPatches = /^a\/(.+).php/g;
+		const patches = await this.getPatches(content, matchPatches);
+		const verRegEx = await this.getVersionRegex(version);
+		const matchHooks = `@since\\s+(${verRegEx})(.*?)(apply_filters|do_action)\\(\\s+(\\'|\\")(.*?)(\\'|\\")`;
```
Comment (roykho) [inline tools/code-analyzer/src/commands/analyzer/index.ts]: Seems this pattern is incomplete. Missing a closing `)`. Also perhaps we should match `do_action( 'hook' );` and without space `do_action('hook');`.

File: tools/code-analyzer/src/commands/analyzer/index.ts [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+				continue;
+			}
+
+			const lines = patch.split('\n');
+			const filepath = await this.getFilename(lines[0]);
+
+			for (const raw of results) {
+				// Extract hook name and type.
+				const hookName = raw.match(
+					/(do_action|apply_filters)\(\s+(\'|\")(.*)(\'|\")/
```
Comment (roykho) [inline tools/code-analyzer/src/commands/analyzer/index.ts]: Same comment as above.

**Review by roykho — CHANGES_REQUESTED** [review]
This is the initial pass of the review. I have left some comments. I also noticed that if you pushed up changes that reverted the change, this tool is still picking up as there are changes. Perhaps we should account for that.

**Comment by roykho** [comment]
So I am still seeing the issue with the reverted code being picked up.

**Comment by roykho** [comment]
I have found several issues and will break it down into two sections. One for CLI and one for GitHub.

**CLI**
* When no changes are found. I am still seeing a lingering header `## HOOKS:`
```
Generating patch for test/pr-cli-pr-review... done
Getting WooCommerce version... done
WooCommerce Version: 6.4.0
Scanning template changes... done
Scanning for new hooks... done
No template changes found

## HOOKS:
```
* When adding an action hook like `do_action()`, I see it being called a filter.
```
NOTICE | New filter found | \'woo_test\' introduced in 6.4.0
```
Perhaps change the `New filter found` to `New hook found`. So that way it can work for both filters and actions.
Second issue there is I am seeing the escape slashes.

**GITHUB**
* When no changes are found, a comment is created with the lingering header `## HOOKS:` just like in the CLI.
* For every change in the PR, a new comment is created. I would suggest to delete all the comments created by this action and add a new one on each trigger. This way you always have one up-to-date comment at the end.

**Review by roykho — CHANGES_REQUESTED** [review]
Left comments.

**Comment by psealock** [comment]
**Status update** - I'm able to reproduce issues found by @roykho and fixed some eslint errors

### Fixed issues
* lingering header `## HOOKS` is resolved

### Unresolved issues
- [x] hooks being called filters in `New filter found`
- [x] Reverted code still produces a Comment
- [x] A new comment is created on every `push` instead of editing or deleting the old
- [x] (**NEW**) Two hooks Introduced in the same file only produces on Notice Fixed in https://github.com/woocommerce/woocommerce/pull/32063/commits/7066bd3997e9c1224a8716f37f3abbdb7625c7ed

**Comment by psealock** [comment]
@roykho Can you give this another look?

## Changes

* Updated RegEx so multiple hooks introduced in the same file can be found.
* Changed `diff` logic to pull down branches to compare instead of downloading a patch from Github. This means removed hooks due to reverted code don't produce a hook result.
* Updated the comment creation to include Update, Delete, and Add. This means there is a single comment being revised/deleted as necessary when hooks are added/updated/removed.

## Testing Instructions

1. Create a branch from this one
2. Add some test hooks. Try multiple hooks, some in the same file
3. Commit changes, push to origin and create a test pull request like [this one](https://github.com/woocommerce/woocommerce/pull/32955).
4. See the comment being added by woocommercebot.
5. Update or remove one hook, commit, push and see the comment updated to reflect changes.
6. Remove all hooks, commit, push and see the comment removed

File: .github/workflows/pr-highlight-changes.yml [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+on: pull_request
+jobs:
+    analyze:
+        name: Check pull request changes to highlight
+        runs-on: ubuntu-20.04
+        outputs:
+            results: ${{ steps.results.outputs.results }}
+        steps:
+            - name: Checkout code
+              uses: actions/checkout@v2
```
Comment (roykho) [inline .github/workflows/pr-highlight-changes.yml]: Let's use v3 to be consistent with all other workflows.

**Review by roykho — CHANGES_REQUESTED** [review]
So in the CLI tool, I am still seeing slashes.
```
FILE: /plugins/woocommerce/woocommerce.php
---------------------------------------------------
HOOK: new_test
---------------------------------------------------
 NOTICE | New action found | \'new_test\' introduced in 6.6.0
---------------------------------------------------
HOOK: woo_test_filter
---------------------------------------------------
 NOTICE | New filter found | \'woo_test_filter\' introduced in 6.6.0
```

Also it would be good if the woocommercebot can add the `release: highlight` label to the PR when there are hook or template changes because we'll need this label to know what needs to be included in the release blog. Also make sure we remove this label if changes are reverted just like how to you remove the woocommercebot comment.

**Comment by psealock** [comment]
Thanks for the review @roykho 

I fixed the `@actions/checkout` to `v3` and removed extra slashes. 

```
## HOOKS:
FILE: /plugins/woocommerce/includes/admin/class-wc-admin-assets.php
---------------------------------------------------
HOOK: woocommerce_test_hook_one
---------------------------------------------------
 NOTICE | New action found | 'woocommerce_test_hook_one' introduced in 6.6.0
---------------------------------------------------
FILE: /plugins/woocommerce/includes/admin/class-wc-admin-attributes.php
---------------------------------------------------
HOOK: woocommerce_test_hook_three
---------------------------------------------------
 NOTICE | New action found | 'woocommerce_test_hook_three' introduced in 6.6.0
---------------------------------------------------
```

I also add and remove the `release: highlight` label depending on if any hooks/template changes have been found. Could you give this another look?

**Comment by roykho** [comment]
@psealock - so now I am seeing an error when I add a hook seen here https://github.com/woocommerce/woocommerce/runs/6406607045?check_suite_focus=true#step:3:14

My testing PR is here https://github.com/woocommerce/woocommerce/pull/32969

**Comment by psealock** [comment]
Darn! Those slashes were needed in the Github context but not the CLI context. https://github.com/woocommerce/woocommerce/pull/32063/commits/5f68f1ed6878cf92fe5090aa617e7f38d3426865 makes a change to the output depending on the context to solve this problem. It should be working now if you can give it another look. Thanks @roykho

**Review by roykho — APPROVED** [review]
LGTM! Tested nicely now. Good Job!

**Comment by roykho** [comment]
Hi @psealock - so just rereading the original task description and actually we also need to detect database changes.

**Comment by psealock** [comment]
Thanks for the review @roykho. I created a new issue to track database changes https://github.com/woocommerce/platform-private/issues/51. I will continue with that next

**Pull request merged**